File: Everybody Codes/2024/d02.rb
DAY,_,_ = __FILE__.rpartition(?.)

require'prime'
require'z3'

r=[]

File.open("#{DAY}a.txt", "r") { |f|
  f.each_line {|l|
    l=l.chomp    
    r+=[l]
  }
}

a,_,s=r

a=a[6..].split(/,/)
p s.scan(/(?=#{a*?|})/).size


# -----------------------


r=[]

File.open("#{DAY}b.txt", "r") { |f|
  f.each_line {|l|
    l=l.chomp
    r+=[l]
  }
}

a,_,*l=r

a=a[6..].split(/,/)
a+=a.map(&:reverse)
p l.sum{|s|
  x=' '*s.size
  a.map{|w|
    s.scan(/(?=#{w})/){v=w.size;x[$`.size,v]='.'*v}
  }
  x.count(?.)
}

# -----------------------

r=[]

File.open("#{DAY}c.txt", "r") { |f|
  f.each_line {|l|
    l=l.chomp
    r+=[l]
    
  }
}

a,_,*l=r

a=a[6..].split(/,/)
a+=a.map(&:reverse)

W,H=l.size,l[0].size

l=l.map{_1*3}
m=l.map{' '*_1.size}
2.times{
  m=l.zip(m).map{|s,x|
    a.map{|w|
      s.scan(/(?=#{w})/){v=w.size;x[$`.size,v]='.'*v}
    }
    x
  }
  l=l.map(&:chars).transpose.map{_1*''}
  m=m.map(&:chars).transpose.map{_1*""}
}
m=m.map{_1[H,H]}
p m.sum{_1.count ?.}

File: flipflop.slome.org/2026/d02.py
DAY,_,_=__file__.rpartition('.')

s=''
with open(f'{DAY}.txt', 'r') as F:
    for l in F:
        l=l.strip()
        s+=l


w=[0]*100
i=0
for c in s:
    if c=='>':i+=1
    else:i-=1
    i%=100
    w[i]+=1

i=w.index(max(w))+1
print(i*max(w))


w=[0]*100
i=j=0
r=0
for c,C in zip(s,s[::-1]):
    if c=='>':i+=1
    else:i-=1
    i%=100
    if C=='>':j+=1
    else:j-=1
    j%=100
    if i==j:r+=1
    else:    w[i]+=1

print(r)